core/trino-main/src/main/java/io/trino/metadata/MetadataListing.java

```diff
@@ -132,31 +135,65 @@ private MetadataListing() {}
 
     public static Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(Session session, Metadata metadata, AccessControl accessControl, QualifiedTablePrefix prefix)
     {
-        Map<SchemaTableName, List<ColumnMetadata>> tableColumns = metadata.listTableColumns(session, prefix).entrySet().stream()
-                .collect(toImmutableMap(entry -> entry.getKey().asSchemaTableName(), Entry::getValue));
+        List<TableColumnsMetadata> catalogColumns = getOnlyElement(metadata.listTableColumns(session, prefix).values(), List.of());
+
+        Map<SchemaTableName, Optional<List<ColumnMetadata>>> tableColumns = catalogColumns.stream()
+                .collect(toImmutableMap(TableColumnsMetadata::getTable, TableColumnsMetadata::getColumns));
+
         Set<SchemaTableName> allowedTables = accessControl.filterTables(
                 session.toSecurityContext(),
                 prefix.getCatalogName(),
                 tableColumns.keySet());
 
         ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> result = ImmutableMap.builder();
-        for (Entry<SchemaTableName, List<ColumnMetadata>> entry : tableColumns.entrySet()) {
-            if (!allowedTables.contains(entry.getKey())) {
-                continue;
+
+        tableColumns.forEach((table, columnsOptional) -> {
+            if (!allowedTables.contains(table)) {
+                return;
             }
-            List<ColumnMetadata> columns = entry.getValue();
+
+            QualifiedObjectName originalTableName = new QualifiedObjectName(prefix.getCatalogName(), table.getSchemaName(), table.getTableName());
+            List<ColumnMetadata> columns;
+
+            if (columnsOptional.isPresent()) {
+                columns = columnsOptional.get();
+            }
+            else {
+                Optional<TableHandle> redirectedTableHandle = Optional.empty();
+                try {
+                    // Handle redirection before filterColumns check
+                    redirectedTableHandle = metadata.getRedirectedTableHandle(session, originalTableName);
+                }
+                catch (TrinoException e) {
+                    if (e.getErrorCode().equals(TABLE_REDIRECTION_ERROR.toErrorCode())) {
```

Comment:
I would invert this, since having an empty "if" branch with an "else" looks funny
```java
// Ignore redirection errors
if (!e.getErrorCode().equals(TABLE_REDIRECTION_ERROR.toErrorCode())) {
    throw e;
}